File: Hard/1425. Constrained Subsequence Sum/1425. Constrained Subsequence Sum.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var constrainedSubsetSum = function (nums, k) {
    const n = nums.length;
    const dp = new Array(n);
    let maxSum = nums[0];

    const deque = [];

    for (let i = 0; i < n; i++) {
        // Calculate the maximum value between 0 and the sum of previous values
        nums[i] += deque.length > 0 ? Math.max(0, dp[deque[0]]) : 0;

        // Pop elements from the front of the deque that are out of the range k
        while (deque.length > 0 && i - deque[0] >= k) {
            deque.shift();
        }

        // Update maxSum
        maxSum = Math.max(maxSum, nums[i]);

        // Pop elements from the back of the deque that are smaller than the current element
        while (deque.length > 0 && nums[i] >= dp[deque[deque.length - 1]]) {
            deque.pop();
        }

        // Push the current index into the deque
        deque.push(i);

        // Update dp array
        dp[i] = nums[i];
    }

    return maxSum;
};

File: Hard/1425. Constrained Subsequence Sum/1425. Constrained Subsequence Sum.java
import java.util.ArrayDeque;
import java.util.Deque;

class Solution {
    public int constrainedSubsetSum(int[] nums, int k) {
        int n = nums.length;
        int[] dp = new int[n];
        int maxSum = nums[0];
        
        Deque<Integer> deque = new ArrayDeque<>();
        
        for (int i = 0; i < n; i++) {
            // Calculate the maximum value between 0 and the sum of previous values
            nums[i] += (deque.isEmpty()) ? 0 : Math.max(0, dp[deque.peekFirst()]);
            
            // Pop elements from the front of the deque that are out of the range k
            while (!deque.isEmpty() && i - deque.peekFirst() >= k) {
                deque.pollFirst();
            }
            
            // Update maxSum
            maxSum = Math.max(maxSum, nums[i]);
            
            // Pop elements from the back of the deque that are smaller than the current element
            while (!deque.isEmpty() && nums[i] >= dp[deque.peekLast()]) {
                deque.pollLast();
            }
            
            // Push the current index into the deque
            deque.offerLast(i);
            
            // Update dp array
            dp[i] = nums[i];
        }
        
        return maxSum;
    }
}
